File: service-go/internal/broker/broker.go
package broker

import (
	"context"
	"fmt"
	"os"

	amqp "github.com/rabbitmq/amqp091-go"
	"github.com/nats-io/nats.go"
	"github.com/segmentio/kafka-go"
)

func envOrDefault(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func Publish(brokerName, target, protocol string, data []byte, origin string) error {
	subject := fmt.Sprintf("serialplab.%s.%s", target, protocol)
	switch brokerName {
	case "kafka":
		return publishKafka(subject, data, origin)
	case "rabbitmq":
		return publishRabbit(subject, data, origin)
	case "nats":
		return publishNats(subject, data, origin)
	default:
		return fmt.Errorf("unknown broker: %s", brokerName)
	}
}

func publishKafka(topic string, data []byte, origin string) error {
	w := &kafka.Writer{
		Addr:  kafka.TCP(envOrDefault("KAFKA_BROKERS", "localhost:11021")),
		Topic: topic,
	}
	defer w.Close()
	return w.WriteMessages(context.Background(), kafka.Message{
		Value: data,
		Headers: []kafka.Header{{Key: "X-Origin", Value: []byte(origin)}},
	})
}

func publishRabbit(routingKey string, data []byte, origin string) error {
	conn, err := amqp.Dial(envOrDefault("RABBITMQ_URL", "amqp://guest:guest@localhost:11022/"))
	if err != nil {
		return err
	}
	defer conn.Close()
	ch, err := conn.Channel()
	if err != nil {
		return err
	}
	defer ch.Close()
	return ch.Publish("amq.topic", routingKey, false, false, amqp.Publishing{
		Body:    data,
		Headers: amqp.Table{"X-Origin": origin},
	})
}

func publishNats(subject string, data []byte, origin string) error {
	nc, err := nats.Connect(envOrDefault("NATS_URL", "nats://localhost:11024"))
	if err != nil {
		return err
	}
	defer nc.Close()
	msg := &nats.Msg{
		Subject: subject,
		Data:    data,
		Header:  nats.Header{"X-Origin": {origin}},
	}
	return nc.PublishMsg(msg)
}

File: service-node/src/broker.ts
import { Kafka } from 'kafkajs';
import amqplib from 'amqplib';
import { connect as natsConnect, headers as natsHeaders } from 'nats';

export async function publish(brokerName: string, target: string, protocol: string, data: Buffer, origin: string): Promise<void> {
  const subject = `serialplab.${target}.${protocol}`;
  switch (brokerName.toLowerCase()) {
    case 'kafka':
      return publishKafka(subject, data, origin);
    case 'rabbitmq':
      return publishRabbit(subject, data, origin);
    case 'nats':
      return publishNats(subject, data, origin);
    default:
      throw new Error(`Unknown broker: ${brokerName}`);
  }
}

async function publishKafka(topic: string, data: Buffer, origin: string): Promise<void> {
  const kafka = new Kafka({ brokers: [process.env.KAFKA_BROKERS || 'localhost:11021'] });
  const producer = kafka.producer();
  await producer.connect();
  await producer.send({ topic, messages: [{ value: data, headers: { 'X-Origin': origin } }] });
  await producer.disconnect();
}

async function publishRabbit(routingKey: string, data: Buffer, origin: string): Promise<void> {
  const conn = await amqplib.connect(process.env.RABBITMQ_URL || 'amqp://guest:guest@localhost:11022');
  const ch = await conn.createChannel();
  ch.publish('amq.topic', routingKey, data, { headers: { 'X-Origin': origin } });
  await ch.close();
  await conn.close();
}

async function publishNats(subject: string, data: Buffer, origin: string): Promise<void> {
  const nc = await natsConnect({ servers: process.env.NATS_URL || 'nats://localhost:11024' });
  const h = natsHeaders();
  h.set('X-Origin', origin);
  nc.publish(subject, data, { headers: h });
  await nc.flush();
  await nc.close();
}
